### ISSUE TITLE
rootDir not including sources under app/assets when compiling tests

### ISSUE THREAD

**Issue by VeryBueno** [joost-de-vries/sbt-typescript#4]
Say that we have the following directory structure, much like your [play-angular2-typesscript](https://github.com/joost-de-vries/play-angular2-typescript) project:
- app/
  - assets/
    - app/
      - `Dog.ts`
- test/
  - assets/
    - `Dog.spec.ts`

If, in `Dog.ts` we have:

``` typescript
export class Dog {
  name:string
}
```

and, in `Dog.spec.ts` we have:

``` typescript
import {Dog} from "../../app/assets/app/Dog"

describe("some test", ()=> {
  it("does a test"), ()=> {
    d:Dog = new Dog();
  }
});
```

I get the following error:

```
[error] tsconfig.json:0: TS6059 File '/projects/myapp/app/assets/app/models/Dog.ts' is not under 'rootDir' '/projects/myapp/test/assets'. 'rootDir' is expected to contain all source files.
```

Is there any way to have the `app/assets/app` directory included when compiling the TypeScript test files?

**Comment by joost-de-vries** [comment]
Which version of sbt-typescript are you using?
It sounds like you may be using an older version.

**Comment by VeryBueno** [comment]
I'm using 0.2.7. I cloned your play-angular2-typescript repo [here](https://github.com/joost-de-vries/play-angular2-typescript) and added the following two files:
- app/assets/app/Dog.ts
- test/assets/Dog.spec.ts

If I run `sbt test:compile`, I get this error:

```
[play-angular2-typescript] $ test:compile
[info] Updating {file:/C:/Projects/pat/}root...
[info] Done updating.
[info] Compiling 8 Scala sources and 1 Java source to C:\Projects\pat\target\scala-2.11\classes...
[info] Typescript linting on 5 source(s)
[info] Typescript compiling on 5 source(s)
[info] Typescript test compiling on 2 source(s)
[info] Typescript test linting on 2 source(s)
[error] fs.js:549
[error]   return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
[error]                  ^
[error]
[error] Error: ENOENT: no such file or directory, open 'C:\Projects\pat\target\web\typescript\test\Dog.spec.js.map'
[error]     at Error (native)
[error]     at Object.fs.openSync (fs.js:549:18)
[error]     at Object.fs.readFileSync (fs.js:397:15)
[error]     at fixSourceMapFile (C:\Projects\pat\project\target\typescript\typescript.js:246:47)
[error]     at C:\Projects\pat\project\target\typescript\typescript.js:225:17
[error]     at Some.map (C:\Projects\pat\project\target\typescript\typescript.js:49:25)
[error]     at C:\Projects\pat\project\target\typescript\typescript.js:215:57
[error]     at Array.map (native)
[error]     at compile (C:\Projects\pat\project\target\typescript\typescript.js:175:71)
[error]     at Object.<anonymous> (C:\Projects\pat\project\target\typescript\typescript.js:147:21)
[info]
[success] Total time: 69 s, completed Jun 8, 2016 8:35:30 AM
```

If I remove `"sourceMap": true` from `tsconfig.json` and run `clean` then `test:compile`, I get this error:

```
[play-angular2-typescript] $ test:compile
[info] Updating {file:/C:/Projects/pat/}root...
[info] Done updating.
[info] Compiling 8 Scala sources and 1 Java source to C:\Projects\pat\target\scala-2.11\classes...
[info] Typescript compiling on 5 source(s)
[info] Typescript linting on 5 source(s)
[info] Typescript test compiling on 2 source(s)
[info] Typescript test linting on 2 source(s)
[error] tsconfig.json:0: TS6059 File 'C:/Projects/pat/app/assets/app/Dog.ts' is not under 'rootDir' 'C:/Projects/pat/test/assets'. 'rootDir' is expected to contain all source files.
[error] one error found
[success] Total time: 77 s, completed Jun 8, 2016 8:41:02 AM
```

Am I doing something wrong? Here's my [Dog.ts](https://github.com/VeryBueno/play-angular2-typescript/blob/master/app/assets/app/Dog.ts) and [Dog.spec.ts](https://github.com/VeryBueno/play-angular2-typescript/blob/master/test/assets/Dog.spec.ts).

I believe this is happening because rootDir is set to `test/assets` when compiling tests and thus any sources in `app/assets/app` are not included. I can see this happening when I add `logLevel in typescript := Level.Debug` in my `build.sbt`:

```
[info] Typescript test compiling on 2 source(s)
[info] Typescript test linting on 2 source(s)
[info] starting compilation of  [ 'Dog.spec.ts', 'todo.spec.ts' ]
[info] from  C:\Projects\pat\test\assets
[info] to  C:\Projects\pat\target\web\typescript\test
[info] args  { sourceFileMappings:
[info]    [ [ 'C:\\Projects\\pat\\test\\assets\\Dog.spec.ts',
[info]        'Dog.spec.ts' ],
[info]      [ 'C:\\Projects\\pat\\test\\assets\\todo.spec.ts',
[info]        'todo.spec.ts' ] ],
[info]   target: 'C:\\Projects\\pat\\target\\web\\typescript\\test',
[info]   options:
[info]    { assetsDir: 'C:\\Projects\\pat\\test\\assets',
[info]      nodeModulesDir: 'C:\\Projects\\pat\\target\\web\\node-modules\\test\\webjars',
[info]      runMode: 'compile',
[info]      tsCodesToIgnore: [],
[info]      extraFiles: [ 'C:\\Projects\\pat\\typings\\index.d.ts' ],
[info]      tsconfigDir: 'C:\\Projects\\pat',
[info]      tsconfig: { compilerOptions: [Object], exclude: [Object] },
[info]      resolveFromNodeModulesDir: true,
[info]      logLevel: 'debug' } }
[info] using tsc options  { outDir: 'C:\\Projects\\pat\\target\\web\\typescript\\test',
[info]   module: 4,
[info]   noImplicitAny: true,
[info]   experimentalDecorators: true,
[info]   noImplicitReturns: true,
[info]   emitDecoratorMetadata: true,
[info]   noFallthroughCasesInSwitch: true,
[info]   target: 1,
[info]   rootDir: 'C:/Projects/pat/test/assets' }
[info] created program
[info] referring to 101 declaration files and 3 code files.
[info] files written [ [ 'C:\\Projects\\pat\\target\\web\\typescript\\test\\Dog.spec.js' ],
[info]   [ 'C:\\Projects\\pat\\target\\web\\typescript\\test\\todo.spec.js' ] ]
[error] tsconfig.json:0: TS6059 File 'C:/Projects/pat/app/assets/app/Dog.ts' is not under 'rootDir' 'C:/Users/Projects/pat/test/assets'. 'rootDir' is expected to contain all source files.
[error] one error found
[success] Total time: 75 s, completed Jun 8, 2016 9:12:57 AM
```

Thank you!

**Comment by joost-de-vries** [comment]
Yeah, I see what's happening.
Unfortunately the current 1.8.x TS compiler supports only one rootDir at this point.

How sbt works is that it turns the main assets dir into a JS module that the test dir refers to.
So I guess it would be possible to not publish the TS sources as build results. And to have the TS compiler produce TS type definitions. Those are allowed to be outside of the root dir.

On the other hand: the 2.0 version of TS is expected to become final this month. And that compiler supports multiple rootdirs. 
So I think it's wiser to fix this in sbt-typescript 3.0snapshot
The snapshot versions of the TS compiler are very stable. F.i. Angular2 already uses it in their RC.

I'll try to make time to fix this this weekend. But a pull request is also welcome.

Tx for pointing this out btw

**Comment by joost-de-vries** [comment]
I've fixed this in the snapshot version of sbt-typescript.
If you get the ts-20-preview branch of play-angular2-typescript it shows that it works. (if you used the snapshot version before you probably should remove it from your ivy cache to get the new version)

**Comment by joost-de-vries** [comment]
Please let me know whether the snapshot version / the ts-20-preview branch of the activator project solves your point.

**Comment by VeryBueno** [comment]
Sorry for the late reply, I've been on vacation for the past few days.

I updated the `projects/plugins.sbt` file in my [play-angular2-typescript](https://github.com/VeryBueno/play-angular2-typescript) clone so that it uses the `0.3.0SNAPSHOT` version of `sbt-typescript`. Then, I did `sbt test:compile` and it did something strange. It created `Dog.spec.js` under `target/web/typescript/test/test/assets/` but it created a _directory_ called Dog.spec.js under `target/web/public/test/public/` but not a JS file. Is this expected? Shouldn't the `Dog.spec.js` be found under `target/web/public/test`?

Also, when I did `sbt clean` and `sbt run` the TypeScript files were not compiled. I received a 404 on `/assets/app/bootstrap.js` when I opened the app in the browser.

Thanks!

**Comment by joost-de-vries** [comment]
Yeah, I've been noticing the same.
The file paths are not resolved correctly even though I'm providing the right rootDirs.
I'll have to find the time to analyse what's going on.

The directories with file names are unfortunate behaviour of sbt-web. They're created when the reported file names don't match the emitted output. Instead of reporting an error. Hmph.
I'm looking into options to make failure more explicit.

**Comment by joost-de-vries** [comment]
Phew. I made a change that fixes [ng2-testdeps](https://github.com/joost-de-vries/sbt-typescript/tree/snapshot-30/src/sbt-test/sbt-typescript/ng2-testdeps) unit test.
So you should be ok now. 
Pls let me know.

**Comment by joost-de-vries** [comment]
The dir instead of file problem still exists unfortunately. But I consider that a different problem. So I'll create a new issue for that.

**Comment by joost-de-vries** [comment]
I just published a new snapshot that improves the output for testassets. But more importantly I added a configuration option `assertCompilation` that performs checks and does extra logging.

This is important since the behaviour of rootDir vs rootDirs is very diferent.
I'm going to register an issue with typescript. The compiler output with rootDirs is so unpredictable that it is really hard to make sbt-typescript robust.
